flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/PhysicalTableSourceScan.scala

```diff
@@ -53,8 +52,9 @@ abstract class PhysicalTableSourceScan(
   protected[flink] val tableSource: TableSource[_] = tableSourceTable.tableSource
 
   override def deriveRowType(): RelDataType = {
-    val flinkTypeFactory = cluster.getTypeFactory.asInstanceOf[FlinkTypeFactory]
-    tableSourceTable.getRowType(flinkTypeFactory)
+    // TableScan row type should always keep same with its
```

Comment:
Not sure about this, if we had project pushed down, will the row type still the same with the table's?

Reply:
If the projection was pushed down, we should modify the relOptTable to keep the row type same as the projected table source. Because the RelOptTable has the ability to convert a table to relational expression, so we should keep it equivalence node with the underlying relational expression.

Reply:
BTW, we already handle the projected fields in TableSourceUtil#getFieldNameType